platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/util/MarkerFileWriter.java

```diff
@@ -0,0 +1,128 @@
+ * directory, no marker files are written.
+ */
+public class MarkerFileWriter {
+
+    /**
+     * The logger for this class.
+     */
+    private static final Logger logger = LogManager.getLogger(MarkerFileWriter.class);
+
+    /**
+     * Flag to indicate if we still need to log an info message about the marker file directory not being set.
+     */
+    private static final AtomicBoolean logMarkerFileDirectoryNotSet = new AtomicBoolean(true);
+
+    /**
+     * Rate limited logger for failure to write marker files.
+     */
+    private final RateLimitedLogger failedToWriteMarkerFileLogger;
+
+    /**
+     * The directory where the marker files are written.  If null, no marker files are written.
+     */
+    private final Path markerFileDirectory;
+
+    /**
+     * Creates a new {@link MarkerFileWriter} with the given {@link PlatformContext}.  If the marker file writer is
+     * enabled, files are written to the configured directory.
+     *
+     * @param platformContext the platform context containing configuration.
+     */
+    public MarkerFileWriter(@NonNull final PlatformContext platformContext) {
+        final boolean enabled = platformContext
+                .getConfiguration()
+                .getConfigData(PathsConfig.class)
+                .writePlatformMarkerFiles();
+        final Path markerFileDirectoryPath = platformContext
+                .getConfiguration()
+                .getConfigData(PathsConfig.class)
+                .getMarkerFilesDir();
+        failedToWriteMarkerFileLogger = new RateLimitedLogger(logger, platformContext.getTime(), Duration.ofMinutes(1));
+        Path directory = null;
+        if (enabled) {
+            if (markerFileDirectoryPath != null) {
+                try {
+                    Files.createDirectories(markerFileDirectoryPath);
+                    directory = markerFileDirectoryPath;
+                } catch (final IOException e) {
+                    directory = null;
+                    logger.error(
+                            LogMarker.EXCEPTION.getMarker(),
+                            "Failed to create marker file directory: {}",
+                            markerFileDirectoryPath,
+                            e);
+                }
+            } else {
+                logger.error(LogMarker.STARTUP.getMarker(), "Marker file writing is turned on but directory is null.");
+            }
+        }
+        markerFileDirectory = directory;
+    }
+
+    /**
+     * Writes a marker file with the given filename to the configured directory.
+     *
+     * @param filename the name of the marker file to write.
+     */
+    public void writeMarkerFile(@NonNull final String filename) {
+        if (markerFileDirectory == null) {
+            // Configuration did not set a marker file directory.  No need to write marker file.
+            return;
+        }
+        final Path markerFile = markerFileDirectory.resolve(filename);
+        if (Files.exists(markerFile)) {
```

Comment:
Since this `MarkerFileWriter` object is being used on the consensus thread, I think we need to ensure that there is no possibility of having any unnecessarily repeated IO calls

What do you think about this solution:
- create a hash set in this class, that stores all `filename`s that have been passed into `writeMarkerFile`
- Before checking `Files.exists` here, add a check to see if the `filename` already exists in the set, and return if so

Reply:
The work of writing marker files is now done inside the context of a `ConcurrentHashMap.computeIfAbsent` check.